### ISSUE TITLE
Add spack/swc to benchmarks once it's ready

### ISSUE THREAD

**Issue by shadowtime2000** [evanw/esbuild#762]
I saw that there was an issue for adding benchmarks for SWC #2, and it was closed because SWC was just a compiler while esbuild is a bundler. There is a tool that is affiliated with SWC in a way called Spack which is built on its same AST structures and all and I think it would be a good idea to add it to the benchmarks.

**Comment by evanw** [comment]
Yeah I’m aware. I try spack every once in a while. Every time so far it has hard-crashed on my benchmark (i.e. panicked). I can check again though. BTW the requirements for the benchmark are to do a minified production build with source maps, so spack will also have to meet that bar.

**Comment by shadowtime2000** [comment]
That makes sense. Not sure about source maps (I think it provides them but idk) but I am pretty sure (based on my usage of `deno bundle` which relies on SWC/Spack) it can minify the output, but it only kills comments and extra whitespace.

**Comment by evanw** [comment]
I don't see a way to have spack do either minification or source maps. The docs say [production mode is currently unimplemented](https://swc.rs/docs/spack-basic#mode):

> Possible values: `production`, `debug`, `none`.
>
> Currently this value is not used, but it will behave similarly to one of webpack.

Doesn't seem like it's ready yet.

**Comment by evanw** [comment]
I'm going to close this since it's not actionable and it's not an issue with esbuild. We can revisit this if/when swc/spack is a production-ready bundler.

**Comment by evanw** [comment]
I have some updates on this. I tried this again recently and it's getting closer to working. You can now run my benchmarks with spack using `make bench-three-spack` and `make bench-rome-spack` in the esbuild repo. Minification configuration options have been added since last time and the recent adjustment to the behavior of spack's `jsc.paths` configuration option made it actually possible to run spack on my TypeScript benchmark.

However, I still don't consider the output of spack to be valid output for these benchmarks, so I'm not going to include it yet. I have observed the following serious problems with the output:

1. Output files are not fully minified, so it's not really a production build. The generated code is bigger than it should be so you'd have to chain another tool after spack to minify it. I'd expect spack to minify everything itself since it already has all of the code for this. I filed an issue about that here: https://github.com/swc-project/swc/issues/2451.

1. The output for the TypeScript benchmark contains a syntax error and crashes when run. The generated code contains a name collision that causes duplicate definition errors. These symbols should have been named separate things. I'm guessing that is this existing issue: https://github.com/swc-project/swc/issues/2717. I was going to wait until that issue is closed before filing my own issue in case they are the same.

These problems would have to be fixed before spack is included in my benchmarks.

It seems like the swc project has a lot of momentum (especially with the [sponsorship from Vercel](https://nextjs.org/blog/next-11-1#adopting-rust-based-swc) and the general excitement about it from Rust enthusiasts) and will likely be a longer-term player in the space, so I'd like to include spack once it generates valid results even if it's not currently as production-ready as the other tools on my benchmarks. Including spack in my benchmarks would ideally help attract testers and contributors and help spack become production-ready sooner than it would have otherwise, which would be great for the community.

Despite the correctness issues, the current performance of spack on my benchmarks looks very promising:

![](https://user-images.githubusercontent.com/406394/141726599-16ef31c1-d2d9-4897-81c1-999aa84ca3fd.png)

I'm observing spack to only be around 10x slower than esbuild on my machine instead of around 100x slower than esbuild like the other tools are. And I don't expect fixes for the bugs I mentioned to have a run-time performance impact so I'd expect the final performance measurements to be around these values.

Reopening this issue since I'm tracking this again.

**Comment by kdy1** [comment]
@evanw Which target did you use? `swc` by default compiles to `es3`, and you have to set `jsc.target` to `es2016` if you want to target `es2015`. (Tracked at https://github.com/swc-project/swc/issues/1108)

As you might know, `es2015` -> `es5` is a huge work.

**Comment by evanw** [comment]
I'm currently using `es2021` as a target since lowering to `es3`/`es5` isn't part of my benchmark:

https://github.com/evanw/esbuild/blob/0f962d694f0ee85f3ec8c8ccd3ec2084d2916dab/Makefile#L493

**Comment by kdy1** [comment]
I instrumented it, and it seems like an issue of transformation, the minifier or the node CLI.
Bundling itself took only `138.047954ms`.

![image](https://user-images.githubusercontent.com/29931815/141867815-cff241d3-814b-41d7-95ee-9d1400c07a42.png)

**Comment by rauchg** [comment]
@evanw I think it'll be crucial that you add the resulting bundle size of each tool, as well as the ability to sort by bundle size or some indication of where each bundler ranks.

One of the reasons we were never able to ship esbuild as a minifier (which would have been a nice incremental adoption path) is that we would have had to make the case of DX at the expense of UX and bandwidth costs.

**Comment by rauchg** [comment]
A proposal would be that we include the output file size next to the timing

<img width="483" alt="CleanShot 2021-11-15 at 17 26 28@2x" src="https://user-images.githubusercontent.com/13041/141878990-6ac434b2-d108-4505-b602-a8ade22c9d91.png">

**Comment by evanw** [comment]
> I instrumented it, and it seems like an issue of transformation, the minifier or the node CLI.

Please let me know if I should be running spack differently to get it to be faster. I was just trying to follow the instructions on the website: https://swc.rs/docs/usage/bundling. There is one exception, which is that I'm calling `node_modules/.bin/spack` directly instead of running `npx spack` since `npx` adds an unavoidable performance penalty (of around 350ms for me).

Note that my benchmark requires the output to be minified. The main point of my benchmark is to test the performance of from-scratch production builds with source maps, which have historically been one of the slowest aspects of JavaScript tooling. Most of the tools on the benchmark (with the exception of esbuild) would be dramatically faster without minification enabled, but that's not the point of the benchmark.

> I think it'll be crucial that you add the resulting bundle size of each tool, as well as the ability to sort by bundle size or some indication of where each bundler ranks.

I didn't list the sizes on the home page because they aren't interesting, but they are already listed on the website in a table here: https://esbuild.github.io/faq/#benchmark-details. They are all around 5.83mb ± 0.04mb so they are basically all the same size and the size differences aren't meaningful IMO. If it was a bar graph it would look something like this:

<img width="738" src="https://user-images.githubusercontent.com/406394/141879857-653f75fb-503d-40bf-9814-79f437784091.png">

Minified size between tools is of course not always so closely-matched as it is in this benchmark. There are a huge number of different dimensions to potentially compare these tools on using more benchmarks. Minification alone has a lot of potential dimensions including supporting partial code evaluation, tree shaking of code nested inside functions, function inlining, object shape modeling, JSX expression hoisting, allocation sinking, method devirtualization, etc. all of which esbuild doesn't currently do but which other tools might do. Then you can get into other aspects of tool choice such as code splitting controls, service worker support, JS/CSS macro systems, HTML entry points, etc. all of which esbuild doesn't currently do either.

But I want the esbuild's home page to just be about bundling performance, since that's the point of the esbuild project, and don't want to dilute the message with a detailed analysis of each tool's benefits and drawbacks. The point of the benchmark on the home page is to show that it's possible to have fast tools for web development, not to help you pick which tool is right for your project. Picking the right tool for your project is a very complex decision. Other websites like https://bundlers.tooling.report/ are a better place for that IMO.

I'm not trying to hide the fact that esbuild has limitations. I try to include them in the documentation to help people make informed decisions. For example, the minification limitations are called out in the documentation for the `minify` option: https://esbuild.github.io/api/#minify-considerations. Here's a copy of what it currently says:

> The minification algorithm in esbuild does not yet do advanced code optimizations. In particular, the following code optimizations are possible for JavaScript code but are not done by esbuild (not an exhaustive list):
>
> * Dead-code elimination within function bodies
> * Function inlining
> * Cross-statement constant propagation
> * Object shape modeling
> * Allocation sinking
> * Method devirtualization
> * Symbolic execution
> * JSX expression hoisting
> * TypeScript enum detection and inlining
>
> If your code makes use of patterns that require some of these forms of code optimization to be compact, or if you are searching for the optimal JavaScript minification algorithm for your use case, you should consider using other tools. Some examples of tools that implement some of these advanced code optimizations include [Terser](https://github.com/terser/terser#readme) and [Google Closure Compiler](https://github.com/google/closure-compiler#readme).

**Comment by evanw** [comment]
Actually now that I'm looking at the bar graph for size, I'm noticing that spack's current output is actually also around the same size as all of the others even though top-level variables are not minified due to https://github.com/swc-project/swc/issues/2451:

<img width="736" src="https://user-images.githubusercontent.com/406394/141884987-77b1255d-5782-4af6-b917-9664dc1a7ace.png">

I was previously looking at the first few lines of spack's output file, which clearly isn't fully minified, but it looks like perhaps top-level variable names aren't a significant size overhead for this particular benchmark. In that case you could argue that spack's output size is acceptable for this benchmark (only 2-4% bigger output seems like it should be considered passing to me).

So I think including spack in my benchmarks is not strictly blocked by https://github.com/swc-project/swc/issues/2451 although it would still be nice to have for production-quality use cases. But I think including spack in my benchmarks should still be blocked by https://github.com/swc-project/swc/issues/2717 since the code in the TypeScript benchmark is still crashing on startup due to duplicate definition errors.

**Comment by kdy1** [comment]
Problem is that `spack` does not support proper minification.
API for minifying is accidentally added to `spack` just because the loader of `spack` uses the same code as `.transform()`.

**Comment by evanw** [comment]
Ah I see, that's good to know. Are you planning on removing the API for minifying from spack in the future? Should I plan on running spack without minification followed by swc with minification in my benchmark instead of just running spack?

From a performance perspective, it'd be ideal if minification happened in the same tool as bundling so that data structures can be reused without re-parsing. But it's also understandable to have them be two separate tools that are run in series instead.

Let me know how I should set this benchmark up. I wouldn't want the benchmark to accidentally pressure you into an architecture direction that you didn't want to go in. Thanks for all of your help with putting this together.

**Comment by kdy1** [comment]
I'll add proper minification config to `spack` and remove minifier configs from `jsc.xxx`. But I'm not sure about the way to benchmark it using CLI at the moment. (I don't think it's possible)

I agree that it should be done within the same tool. I'll improve the bundler in near future, and I think adding it to the list after implementing proper minification config would be better.

**Comment by evanw** [comment]
Sounds good. I'll wait until the minification config is changed then.

I also just saw the note at the top of https://swc.rs/docs/configuration/bundling which says that the name will be changed to swcpack soon and that the configuration file is going to change too:

> This feature is currently named `spack`, but will be renamed to `swcpack` in `v2`. `spack.config.js` will be deprecated for `swcpack.config.js`.

So I should also wait until the rename happens before adding spack to my benchmarks to avoid causing confusion.

**Comment by evanw** [comment]
Closing for now since this still isn't actionable.
